### PR TITLE
Add implementation for foreign-object

### PR DESCRIPTION
**Pull request by f-f**
This PR adds FFI for foreign-object using a symbol-hashtable together with a list to keep track of insertion order, which is an invariant of the JS implementation.

It's not strictly needed for Scheme, but it's good to have for compatibility with existing JS-backend code.

There is one slow test that I need to look more into, but the patch passes all tests otherwise.

### CODE REVIEW

**Comment by f-f** [comment]
CI fails in a strange way:
```
[1/2 ModuleNotFound] foreign-object/src/Foreign/Object.purs:63:1

  63  import Type.Row.Homogeneous (class Homogeneous)
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  Module Type.Row.Homogeneous was not found.
  Make sure the source file exists, and that it has been provided as an input to the compiler.

[2/2 ModuleNotFound] foreign-object/test/Test/Foreign/Object.purs:13:1

  13  import Data.Map as M
      ^^^^^^^^^^^^^^^^^^^^

  Module Data.Map was not found.
  Make sure the source file exists, and that it has been provided as an input to the compiler.
```

These modules are supposed to be there.

File: foreign-object/src/Foreign/Object.ss [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+          keys)
+  (import (only (rnrs base) define lambda if let let* let-values begin)
+          (prefix (chezscheme) scm:)
+          (prefix (purs runtime) rt:)
+          (prefix (purs runtime srfi :214) arrays:)
+          (only (purs runtime pstring) pstring->symbol string->pstring))
+
+  (define symbol->pstring
+    (lambda (s)
+      (string->pstring (scm:symbol->string s))))
```
Comment (anttih) [inline foreign-object/src/Foreign/Object.ss]: This has been upstreamed to `(purs runtime pstring)`.

File: foreign-object/src/Foreign/Object.ss [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+          _foldSCObject
+          all
+          size
+          _lookup
+          _lookupST
+          fromHomogeneousImpl
+          toArrayWithKey
+          keys)
+  (import (only (rnrs base) define lambda if let let* let-values begin)
+          (prefix (chezscheme) scm:)
```
Comment (anttih) [inline foreign-object/src/Foreign/Object.ss:19]: I've imported this unqualified, to make the code a bit more terse, but it's fine if you want to qualify.

**Review by anttih — COMMENTED** [review]
How about using an alist in a `box` for the `Object`? That way you can approximate a mutable object but still be faster than hashtable.

**Comment by f-f** [comment]
> How about using an alist in a box for the Object? That way you can approximate a mutable object but still be faster than hashtable.

I tried the alist but without the box and had issues with the mutability, didn't try wrapping in a box. Are hashtables really that slow? I'll try this out

**Comment by anttih** [comment]
You need to have a lot of entries (like ~hundreds) until hashtables are faster than alists, so in that sense they are slow.

**Comment by f-f** [comment]
@anttih moved to alist in a box, it's much faster now.
CI fails because of https://github.com/purescm/purescm/pull/244

File: foreign-object/src/Foreign/Object.ss [hunk 3]
```diff
@@ -19,141 +19,120 @@
...
-        (let ([ks1 (scm:reverse (scm:cdr m))])
-          (let loop ([ks ks1]
-                    [vs (scm:map (lambda (k) (scm:symbol-hashtable-ref (scm:car m) k #f)) ks1)]
-                    [i 0]
-                    [acc (arrays:make-flexvector (scm:length ks1))])
-            (if (scm:null? ks)
-              acc
+        (let* ([len (scm:length (scm:unbox m))]
+              [arr (arrays:make-flexvector len)])
+          (let loop ([i (scm:- len 1)]
```
Comment (anttih) [inline foreign-object/src/Foreign/Object.ss]: Could use `fx-` here and below.

**Review by anttih — APPROVED** [review]

**Comment by f-f** [comment]
CI is stuck because the latest Spago is broken, see https://github.com/purescript/spago/issues/1242

**Pull request merged**